odl/util/vectorization.py

```diff
@@ -124,169 +125,240 @@ def out_shape_from_array(arr):
-
-            order = meshgrid_input_order(x)
-            vecs = vecs_from_meshgrid(x, order=order)
-            if ndim == 1:
-                for i, pt in enumerate(vecs[0]):
-                    out[i] = func(pt, **kwargs)
-            elif dtype is None:
-                for i, pt in enumerate(product(*vecs)):
-                    # Worst multi-index implementation ever, but still good
-                    # enough for slow code
-                    index = np.unravel_index(i, out_shape)
-                    sub = out
-                    for j in index[:-1]:
-                        sub = sub[j]
-                    sub[index[-1]] = func(pt, **kwargs)
-            else:
-                for i, pt in enumerate(product(*vecs)):
-                    out.flat[i] = func(pt, **kwargs)
-            return np.asarray(out)
-
-        @wraps(func)
-        def vect_wrapper(x, out=None, **kwargs):
-            # Find out dimension first
-            if isinstance(x, np.ndarray):  # array
-                if x.ndim == 1:
-                    ndim = 1
-                elif x.ndim == 2:
-                    ndim = len(x)
-                else:
-                    raise ValueError('only 1- or 2-dimensional arrays '
-                                     'supported.')
-            else:  # meshgrid or single value
-                try:
-                    ndim = len(x)
-                except TypeError:
-                    ndim = 1
-
-            if is_valid_input_meshgrid(x, ndim):
-                return _vect_wrapper_meshgrid(x, out, ndim, **kwargs)
-            elif is_valid_input_array(x, ndim):
-                return _vect_wrapper_array(x, out, **kwargs)
-            else:
-                try:
-                    return func(x)
-                except Exception as err:
-                    raise_from(
-                        TypeError('invalid vectorized input type.'), err)
-
-        return vect_wrapper
-    return vect_decorator
+    def decorator(func):
+        if not use_kw:
+            def wrapper(x):
+                if len(x) != nin:
+                    raise TypeError('Need a sequence of length {} as input, '
+                                    'got length {}.'.format(nin, len(x)))
+                return func(*x)
+        else:
+            def wrapper(x, **kwargs):
+                if len(x) != nin:
+                    raise TypeError('Need a sequence of length {} as input, '
+                                    'got length {}.'.format(nin, len(x)))
+                return func(*x, **kwargs)
+
+        return wrapper
+    return decorator
+
+
+class vectorize(object):
+
+    """Decorator class for function vectorization.
+
+    This vectorizer expects a function with exactly one positional
+    argument (input) and optional keyword arguments. The decorated
+    function has an optional ``out`` parameter for in-place evaluation.
+    """
+
+    def __new__(cls, *args, **kwargs):
```

Comment:
I don't think there is anything standard doing this, but I haven't searched very intensely either. This functionality could also be implemented with the help of the [wrapt](https://pypi.python.org/pypi/wrapt) module, but I find its syntax slightly confusing, and we don't need its full power.

I'll factor out the pattern.